### PR TITLE
allow JWT algorithms restrictions

### LINKED ISSUE #2714: Add a configurable restriction of algorithms per issuer
we should be able to define the set of accepted algorithms for an issuer when validating a JWT

### PR DESCRIPTION
**Pull request by Geal**
Fix #2714
Issuers can provide a JWKS containing keys usable with multiple algorithms, so we might want to restrict the list of accepted algorithms to a well known set

**Checklist**

Complete the checklist (and note appropriate exceptions) before a final PR is raised.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]. It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]. Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]. Tick whichever testing boxes are applicable. If you are adding Manual Tests:
    - please document the manual testing (extensively) in the Exceptions.
    - please raise a separate issue to automate the test and label it (or ask for it to be labeled) as `manual test`

### CODE REVIEW

File: apollo-router/src/plugins/authentication/mod.rs [hunk 1]
```diff
@@ -120,6 +120,10 @@ struct JwksConf {
     url: String,
     /// Expected issuer for tokens verified by that JWKS
     issuer: Option<String>,
+    /// List of accepted algorithms. Possible values are HS2356, HS384, HS512, ES256, ES384, RS256, RS384, RS512, PS256, PS384, PS512, EdDSA
```
Comment (garypen) [inline apollo-router/src/plugins/authentication/mod.rs:123]: I don't have a solution, but we'll have to remember to keep this up to date if the crate adds/remove algorithms. One option, don't list all the possibilities, but provide a reference to the crate.

File: apollo-router/src/plugins/authentication/mod.rs [hunk 2]
```diff
@@ -167,7 +171,17 @@ fn search_jwks(
...
     const HIGHEST_SCORE: usize = 2;
     let mut candidates = vec![];
-    for JwkSetInfo { jwks, issuer } in jwks_manager.iter_jwks() {
+    for JwkSetInfo {
+        jwks,
+        issuer,
+        algorithms,
+    } in jwks_manager.iter_jwks()
+    {
+        // filter accepted algorithms
```
Comment (garypen) [inline apollo-router/src/plugins/authentication/mod.rs:180]: NIT: Do we really need the is_empty() check first? If algorithms is a `HashSet` I would expect contains to be fast...

File: docs/source/configuration/authn-jwt.mdx [hunk 3]
```diff
@@ -94,6 +94,7 @@ The following configuration options are supported:
   - **If you use a third-party IdP,** consult its documentation to determine its JWKS URL.
   - **If you use your own custom IdP,** you need to make its JWKS available at a router-accessible URL if you haven't already. For more information, see [Creating your own JWKS](#creating-your-own-jwks-advanced).
 - `issuer`: **optional** name of the issuer, that will be compared to the `iss` claim in the JWT if present. If it does not match, the request will be rejected.
+- `algorithms`: **optional** list of accepted algorithms. Possible values are HS2356, HS384, HS512, ES256, ES384, RS256, RS384, RS512, PS256, PS384, PS512, EdDSA
```
Comment (garypen) [inline docs/source/configuration/authn-jwt.mdx:97]: Same comment as above.

File: apollo-router/src/plugins/authentication/mod.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/plugins/authentication/mod.rs:123]: Is empty is the case where no filtering is set. Would you prefer an Option there?

File: apollo-router/src/plugins/authentication/mod.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/src/plugins/authentication/mod.rs:180]: Is empty is the case where no filtering is set. Would you prefer an Option there?

File: docs/source/configuration/authn-jwt.mdx (hunk 3 above)
Comment (StephenBarlow) [inline docs/source/configuration/authn-jwt.mdx:97]: Could we put each of those individual supported values in `codefont`

File: apollo-router/src/plugins/authentication/mod.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/plugins/authentication/mod.rs:123]: The list of algorithms is defined in an IETF RFC, it does not change often

File: apollo-router/src/plugins/authentication/mod.rs (hunk 2 above)
Comment (Geal) [reply apollo-router/src/plugins/authentication/mod.rs:180]: moved to an option in [965f4ad](https://github.com/apollographql/router/pull/2852/commits/965f4adf84dedb6f014bc241b2ebcb1ce04ebafd)

**Review by bnjjj — APPROVED** [review]

**Pull request merged**